### PR TITLE
initial commit for cloudrun build and deploy workflows

### PR DESCRIPTION
**Pull request by cru-Luis-Rodriguez**
## Build & Deploy Cloud Run Function Workflow

This repository contains a GitHub Actions workflow for building and deploying a Cloud Function to Google Cloud Run. The workflow is defined in `workflow-templates/build-deploy-cloudrun-function.yml`.

### Workflow Configuration

The workflow is triggered on:
- Push events to the `$default-branch`.
- Manual dispatch via the GitHub Actions interface.

#### Inputs

- `function_name`: The name of the Cloud Function to deploy (required).
- `entry_point`: The Python function serving as the entry point (required).
- `runtime`: The function runtime (required).

#### Secrets

The following secrets must be configured in your GitHub repository:
- `GCP_PROJECT_ID`
- `GCP_PROJECT_NUMBER`
- `WORKLOAD_IDENTITY_POOL`
- `WORKLOAD_IDENTITY_PROVIDER`
- `GCP_SERVICE_ACCOUNT`
- `GCP_SERVICE_ACCOUNT_EMAIL`
- `GCP_REGION`

## Notes

- Ensure that all required secrets are configured in your GitHub repository.
- Update the paths and inputs as necessary to match your project structure and requirements.

### CODE REVIEW

**Comment by mattdrees** [comment]
Cool, thanks for working on this. I'd be happy to have you own the GHA deployment stuff.

On the plane I was starting to work on a gha workflow, but I didn't think of putting in our org .github directory. I will have to research when it's appropriate to put things here vs the repo itself.

I think the flow of what you have here is super close to what I was drafting.
Main difference is that mine doesn't do docker building; I let Cloud Functions do that part internally. Mine pretty much just calls 'glcoud functions deploy'. I'll commit mine in a draft PR in the other repo and point you toward it, for comparison's sake.

**Comment by mattdrees** [comment]
Here's my WIP draft:
https://github.com/CruGlobal/dot/pull/1/commits/2ba57e0246cd07059feeaa84fffa1fb6450a366b

**Comment by cru-Luis-Rodriguez** [comment]
@mattdrees As far as where to store the build and deploy workflow. I lean toward setting it here so that other repos can use it without duplicating the process. 
I've incorporated your deployment process and opted to use the gcloud functions deployment. I'll refactor the build-deploy-cloudrun-job.yml so that it is obvious it uses the docker build. That way, we can have both options.

File: .github/workflows/build-deploy-cloudrun-function.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        type: string
+        required: true
+      GCP_SERVICE_ACCOUNT_EMAIL:
+        description: 'GCP Service Account Email'
+        type: string
+        required: true
+      GCP_REGION:
+        description: 'GCP Region'
+        type: string
+        required: true
```
Comment (Omicron7) [inline .github/workflows/build-deploy-cloudrun-function.yml:27]:
```suggestion
  workflow_call:
    # Allows manual build and deploy of any branch/ref
    inputs:
      function_name:
        description: 'The name of the Cloud Function to deploy'
        type: string
        required: true
    secrets:
      WORKLOAD_IDENTITY_PROVIDER:
        description: 'Workload Identity Provider'
        type: string
        required: true
      GCP_SERVICE_ACCOUNT_EMAIL:
        description: 'GCP Service Account Email'
        type: string
        required: true
      GCP_REGION:
        description: 'GCP Region'
        type: string
        required: true
```

File: .github/workflows/build-deploy-cloudrun-function.yml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          gcloud config set project ${{ secrets.GCP_PROJECT_ID }}
+          gcloud functions deploy ${{ inputs.function_name }}-${{ needs.build.outputs.build-number }} \
+           --no-allow-unauthenticated \
+            --region $GCP_REGION \
+            --gen2 \
+            --source=./${{ inputs.function_name }} \
+            --runtime=python312 \
+            --entry-point=hello_http \
+            --trigger-http \
+            --set-secrets=API_KEY=${{ inputs.function_name }}_API_KEY:latest,API_SECRET=${{ inputs.function_name }}_API_SECRET:latest \
```
Comment (mattdrees) [inline .github/workflows/build-deploy-cloudrun-function.yml:91]: I would have hoped this line wouldn't be necessary, because it's set up in terraform.

**Review by mattdrees — COMMENTED** [review]
I'll look at it more tonight

File: .github/workflows/build-deploy-cloudrun-function.yml (hunk 2 above)
Comment (cru-Luis-Rodriguez) [reply .github/workflows/build-deploy-cloudrun-function.yml:91]: Where in Terraform is this set? I'm assuming you mean all four lines 88,89,90,91.

Comment (mattdrees) [reply .github/workflows/build-deploy-cloudrun-function.yml:97]: not 88-90; just 91.
It's here: https://github.com/CruGlobal/cru-terraform-modules/blob/cloudrun/gcp/cloudrun-function/scheduled-tasks/main.tf#L72-L80

**Comment by cru-Luis-Rodriguez** [comment]
@mattdrees @Omicron7 ready for review.

File: workflow-templates/build-deploy-cloudrun-function.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      runtime: { runtime } # python312
+      environment: { environment }  # production
+    secrets:
+      GCP_PROJECT_ID: ${{ vars.GCP_PROJECT_ID }}
+      GCP_PROJECT_NUMBER: ${{ vars.GCP_PROJECT_NUMBER }}
+      WORKLOAD_IDENTITY_POOL: ${{ vars.WORKLOAD_IDENTITY_POOL }}
+      WORKLOAD_IDENTITY_PROVIDER: ${{ vars.WORKLOAD_IDENTITY_PROVIDER }}
+      GCP_SERVICE_ACCOUNT: ${{ vars.GCP_SERVICE_ACCOUNT }}
+      GCP_SERVICE_ACCOUNT_EMAIL: ${{ vars.GCP_SERVICE_ACCOUNT_EMAIL }}
+      GCP_REGION: ${{ vars.GCP_REGION }}
```
Comment (mattdrees) [inline workflow-templates/build-deploy-cloudrun-function.yml:41]: Is this `secrets` section necessary?

Comment (mattdrees) [reply workflow-templates/build-deploy-cloudrun-function.yml:41]: Ok, thanks for testing. I had hoped that this was only necessary for secrets and not for configuration variables.

**Review by mattdrees — APPROVED** [review]

**Pull request merged**